Prior Value in Rolling Change: nonces differ

Seeing this IllegalStateException on latest besu 22.7.0-RC1 on goerli canary (fast sync, bonsai, from scratch)

LOGS
```{"@timestamp":"2022-07-06T16:01:56,567","level":"INFO","thread":"EthScheduler-Workers-0","class":"BlockPropagationManager","message":"Imported 1 pending blocks","throwable":""}
{"@timestamp":"2022-07-06T16:02:01,016","level":"INFO","thread":"Timer-0","class":"DNSTimerTask","message":"Refreshing DNS records with enrtree://AKA3AM6LPBYEUDMVNU3BSVQJ5AD45Y7YPOHJLEF6W26QOE4VTUDPE@all.goerli.ethdisco.net","throwable":""}
{"@timestamp":"2022-07-06T16:02:06,309","level":"INFO","thread":"EthScheduler-Workers-1","class":"PersistBlockTask","message":"Imported #7,180,290 / 9 tx / 0 om / 4,224,326 (14.1%) gas / (0xb70f501043b17794d75447fb9996030647151f1fc7c0a28a1ab0a5dcf02c212b) in 0.153s. Peers
: 72","throwable":""}
java.lang.RuntimeException: java.lang.IllegalStateException: Address=0xa550511f2093dbef0e277ec1b216849d4d73bf1a Prior Value in Rolling Change: nonces differ
        at org.hyperledger.besu.ethereum.bonsai.BonsaiWorldStateArchive.getMutable(BonsaiWorldStateArchive.java:254)
        at org.hyperledger.besu.ethereum.MainnetBlockValidator.validateAndProcessBlock(MainnetBlockValidator.java:100)
        at org.hyperledger.besu.ethereum.MainnetBlockValidator.validateAndProcessBlock(MainnetBlockValidator.java:71)
        at org.hyperledger.besu.ethereum.mainnet.MainnetBlockImporter.importBlock(MainnetBlockImporter.java:44)
        at org.hyperledger.besu.ethereum.core.BlockImporter.importBlock(BlockImporter.java:44)
        at org.hyperledger.besu.ethereum.eth.sync.tasks.PersistBlockTask.executeTask(PersistBlockTask.java:201)
        at org.hyperledger.besu.ethereum.eth.manager.task.AbstractEthTask.executeTaskTimed(AbstractEthTask.java:150)
        at org.hyperledger.besu.ethereum.eth.manager.task.AbstractEthTask.run(AbstractEthTask.java:75)
        at org.hyperledger.besu.ethereum.eth.sync.BlockPropagationManager.runImportTask(BlockPropagationManager.java:487)
        at org.hyperledger.besu.ethereum.eth.sync.BlockPropagationManager.validateAndProcessPendingBlock(BlockPropagationManager.java:468)
        at org.hyperledger.besu.ethereum.eth.sync.BlockPropagationManager.lambda$importOrSavePendingBlock$17(BlockPropagationManager.java:456)
        at org.hyperledger.besu.util.FutureUtils.propagateResult(FutureUtils.java:97)
        at org.hyperledger.besu.ethereum.eth.manager.EthScheduler.lambda$scheduleSyncWorkerTask$0(EthScheduler.java:108)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.IllegalStateException: Address=0xa550511f2093dbef0e277ec1b216849d4d73bf1a Prior Value in Rolling Change: nonces differ
        at org.hyperledger.besu.ethereum.bonsai.BonsaiAccount.assertCloseEnoughForDiffing(BonsaiAccount.java:303)
        at org.hyperledger.besu.ethereum.bonsai.BonsaiWorldStateUpdater.rollAccountChange(BonsaiWorldStateUpdater.java:497)
        at org.hyperledger.besu.ethereum.bonsai.BonsaiWorldStateUpdater.lambda$rollBack$13(BonsaiWorldStateUpdater.java:447)
        at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1764)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
        at org.hyperledger.besu.ethereum.bonsai.BonsaiWorldStateUpdater.rollBack(BonsaiWorldStateUpdater.java:445)
        at org.hyperledger.besu.ethereum.bonsai.BonsaiWorldStateArchive.getMutable(BonsaiWorldStateArchive.java:239)
        ... 17 more
{"@timestamp":"2022-07-06T16:02:06,684","level":"ERROR","thread":"EthScheduler-Workers-0","class":"MainnetBlockValidator","message":"Unable to process block because parent world state 0x3c111c8df49432fbcb5cc316548a1178410c521f7806d099eabe35b5920e7ac5 is not available. Block 7180287 (0x651751749dad691a615d650461944b2696a9e7059216266967399792752dd0b1)","throwable":""}
{"@timestamp":"2022-07-06T16:02:06,685","level":"WARN","thread":"EthScheduler-Workers-0","class":"BlockPropagationManager","message":"Failed to import announced block 7180287 (0x651751749dad691a615d650461944b2696a9e7059216266967399792752dd0b1)","throwable":""}
java.lang.RuntimeException: java.lang.IllegalStateException: Address=0xa550511f2093dbef0e277ec1b216849d4d73bf1a Prior Value in Rolling Change: nonces differ
        at org.hyperledger.besu.ethereum.bonsai.BonsaiWorldStateArchive.getMutable(BonsaiWorldStateArchive.java:254)
        at org.hyperledger.besu.ethereum.MainnetBlockValidator.validateAndProcessBlock(MainnetBlockValidator.java:100)
        at org.hyperledger.besu.ethereum.MainnetBlockValidator.validateAndProcessBlock(MainnetBlockValidator.java:71)
        at org.hyperledger.besu.ethereum.mainnet.MainnetBlockImporter.importBlock(MainnetBlockImporter.java:44)
        at org.hyperledger.besu.ethereum.core.BlockImporter.importBlock(BlockImporter.java:44)
        at org.hyperledger.besu.ethereum.eth.sync.tasks.PersistBlockTask.executeTask(PersistBlockTask.java:201)
        at org.hyperledger.besu.ethereum.eth.manager.task.AbstractEthTask.executeTaskTimed(AbstractEthTask.java:150)
        at org.hyperledger.besu.ethereum.eth.manager.task.AbstractEthTask.run(AbstractEthTask.java:75)
        at org.hyperledger.besu.ethereum.eth.sync.BlockPropagationManager.runImportTask(BlockPropagationManager.java:487)
        at org.hyperledger.besu.ethereum.eth.sync.BlockPropagationManager.validateAndProcessPendingBlock(BlockPropagationManager.java:468)
        at org.hyperledger.besu.ethereum.eth.sync.BlockPropagationManager.lambda$importOrSavePendingBlock$17(BlockPropagationManager.java:456)
        at org.hyperledger.besu.util.FutureUtils.propagateResult(FutureUtils.java:97)
        at org.hyperledger.besu.ethereum.eth.manager.EthScheduler.lambda$scheduleSyncWorkerTask$0(EthScheduler.java:108)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

Node does not recover
```
        ... 17 more
{"@timestamp":"2022-07-06T16:02:16,063","level":"ERROR","thread":"EthScheduler-Workers-3","class":"MainnetBlockValidator","message":"Unable to process block because parent world state 0xe80ffa90064b2135e93a06afaefba7ffae7554e77ce3d5f4e64150f7f5f320af is not available. Blo
ck 7180285 (0x0868c83578e6eac93e138f6ba27bbbb88fa295fe2f0efb9534308332d581fbda)","throwable":""}
{"@timestamp":"2022-07-06T16:02:16,063","level":"WARN","thread":"EthScheduler-Workers-3","class":"BlockPropagationManager","message":"Failed to import announced block 7180285 (0x0868c83578e6eac93e138f6ba27bbbb88fa295fe2f0efb9534308332d581fbda)","throwable":""}
java.lang.RuntimeException: java.lang.IllegalStateException: Address=0xa550511f2093dbef0e277ec1b216849d4d73bf1a Prior Value in Rolling Change: nonces differ
```
And eventually there's an invalid block
```

        ... 13 more
{"@timestamp":"2022-07-07T05:01:51,989","level":"ERROR","thread":"EthScheduler-Services-183 (importBlock)","class":"MainnetBlockValidator","message":"Unable to process block because parent world state 0x0ef0262e5aca01ca9dbf1281a1e189a25dca4af1ad8d390a5ca197cd9ec5e0e9 is not available. Block 7180291 (0xc9efc909fcf9ba5a34653cb0e2e07bbaaf4d61b539fd1cb9287bb1ae759f1c39)","throwable":""}
{"@timestamp":"2022-07-07T05:01:51,990","level":"WARN","thread":"EthScheduler-Services-183 (importBlock)","class":"PipelineChainDownloader","message":"Invalid block detected. Disconnecting from sync target. Failed to import block: Invalid block at #7180291 (0xc9efc909fcf9ba5a34653cb0e2e07bbaaf4d61b539fd1cb9287bb1ae759f1c39)","throwable":""}
{"@timestamp":"2022-07-07T05:01:51,990","level":"WARN","thread":"EthScheduler-Services-183 (importBlock)","class":"PipelineChainDownloader","message":"Chain download failed. Restarting after short delay.","throwable":" java.util.concurrent.CompletionException: org.hyperledger.besu.ethereum.eth.sync.tasks.exceptions.InvalidBlockException: Failed to import block: Invalid block at #7180291 (0xc9efc909fcf9ba5a34653cb0e2e07bbaaf4d61b539fd1cb9287bb1ae759f1c39)\n\tat java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:367)\n\tat java.base/java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:376)\n\tat java.base/java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:1019)\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)\n\tat java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)\n\tat org.hyperledger.besu.services.pipeline.Pipeline.abort(Pipeline.java:180)\n\tat org.hyperledger.besu.services.pipeline.Pipeline.lambda$runWithErrorHandling$3(Pipeline.java:159)\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat java.base/java.lang.Thread.run(Thread.java:829)\nCaused by: org.hyperledger.besu.ethereum.eth.sync.tasks.exceptions.InvalidBlockException: Failed to import block: Invalid block at #7180291 (0xc9efc909fcf9ba5a34653cb0e2e07bbaaf4d61b539fd1cb9287bb1ae759f1c39)\n\tat org.hyperledger.besu.ethereum.eth.sync.fullsync.FullImportBlockStep.accept(FullImportBlockStep.java:62)\n\tat org.hyperledger.besu.ethereum.eth.sync.fullsync.FullImportBlockStep.accept(FullImportBlockStep.java:31)\n\tat org.hyperledger.besu.services.pipeline.CompleterStage.run(CompleterStage.java:37)\n\tat org.hyperledger.besu.services.pipeline.Pipeline.lambda$runWithErrorHandling$3(Pipeline.java:152)\n\t... 5 more\n"}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,7 +11,6 @@
 - When on PoS the head can be only be updated by ForkchoiceUpdate [#3994](https://github.com/hyperledger/besu/pull/3994)
 - Version information available in metrics [#3997](https://github.com/hyperledger/besu/pull/3997)
 - Add TTD and DNS to Sepolia config [#4024](https://github.com/hyperledger/besu/pull/4024)
-- Add terminal block hash and number to Ropsten genesis file [#4026](https://github.com/hyperledger/besu/pull/4026)
 - Return `type` with value `0x0` when serializing legacy transactions [#4027](https://github.com/hyperledger/besu/pull/4027)
 - Ignore `ForkchoiceUpdate` if `newHead` is an ancestor of the chain head [#4055](https://github.com/hyperledger/besu/pull/4055)
 
@@ -20,6 +19,7 @@
 - Support free gas networks in the London fee market [#4003](https://github.com/hyperledger/besu/pull/4003)
 - Limit the size of outgoing eth subprotocol messages.  [#4034](https://github.com/hyperledger/besu/pull/4034)
 - Fixed a state root mismatch issue on bonsai that may appear occasionally [#4041](https://github.com/hyperledger/besu/pull/4041)
+- Fixed a trie log layer issue on bonsai during reorg [#4069](https://github.com/hyperledger/besu/pull/4069)
 - Return the correct latest valid hash in case of bad block when calling engine methods [#4056](https://github.com/hyperledger/besu/pull/4056)
 
 ### Download links

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,9 @@
 
 ### Additions and Improvements
 
+### Bug Fixes
+- Fixed a trie log layer issue on bonsai during reorg [#4069](https://github.com/hyperledger/besu/pull/4069)
+
 ## 22.7.0-RC1
 
 ### Additions and Improvements
@@ -19,7 +22,6 @@
 - Support free gas networks in the London fee market [#4003](https://github.com/hyperledger/besu/pull/4003)
 - Limit the size of outgoing eth subprotocol messages.  [#4034](https://github.com/hyperledger/besu/pull/4034)
 - Fixed a state root mismatch issue on bonsai that may appear occasionally [#4041](https://github.com/hyperledger/besu/pull/4041)
-- Fixed a trie log layer issue on bonsai during reorg [#4069](https://github.com/hyperledger/besu/pull/4069)
 - Return the correct latest valid hash in case of bad block when calling engine methods [#4056](https://github.com/hyperledger/besu/pull/4056)
 
 ### Download links

--- a/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiPersistedWorldState.java
+++ b/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiPersistedWorldState.java
@@ -252,9 +252,17 @@ public class BonsaiPersistedWorldState implements MutableWorldState, BonsaiWorld
       // then persist the TrieLog for that transition.
       // If specified but not a direct descendant simply store the new block hash.
       if (blockHeader != null) {
-        final TrieLogLayer trieLog =
-            prepareTrieLog(blockHeader, localUpdater, newWorldStateRootHash);
-        persistTrieLog(blockHeader, newWorldStateRootHash, trieLog, stateUpdater);
+        // do not overwrite a trielog layer that already exists in the database.
+        // if it's only in memory we need to save it
+        // for example, like that in case of reorg we don't replace a trielog layer
+        if (worldStateStorage.getTrieLog(blockHeader.getHash()).isEmpty()) {
+          final TrieLogLayer trieLog =
+              prepareTrieLog(blockHeader, localUpdater, newWorldStateRootHash);
+          persistTrieLog(blockHeader, newWorldStateRootHash, trieLog, stateUpdater);
+        }
+        stateUpdater
+            .getTrieBranchStorageTransaction()
+            .put(WORLD_BLOCK_HASH_KEY, blockHeader.getHash().toArrayUnsafe());
         worldStateBlockHash = blockHeader.getHash();
       } else {
         stateUpdater.getTrieBranchStorageTransaction().remove(WORLD_BLOCK_HASH_KEY);
@@ -310,9 +318,6 @@ public class BonsaiPersistedWorldState implements MutableWorldState, BonsaiWorld
         "Persisting trie log for block hash {} and world state root {}",
         blockHeader::toLogString,
         worldStateRootHash::toHexString);
-    stateUpdater
-        .getTrieBranchStorageTransaction()
-        .put(WORLD_BLOCK_HASH_KEY, blockHeader.getHash().toArrayUnsafe());
     final BytesValueRLPOutput rlpLog = new BytesValueRLPOutput();
     trieLog.writeTo(rlpLog);
     stateUpdater

--- a/ethereum/core/src/test/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchiveTest.java
+++ b/ethereum/core/src/test/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchiveTest.java
@@ -229,16 +229,10 @@ public class BonsaiWorldStateArchiveTest {
     // initial persisted state hash key
     when(blockchain.getBlockHeader(eq(Hash.ZERO))).thenReturn(Optional.of(blockHeaderChainA));
     // fake trie log layer
-    final BytesValueRLPOutput rlpLogBlockA = new BytesValueRLPOutput();
-    final TrieLogLayer trieLogLayerBlockA = new TrieLogLayer();
-    trieLogLayerBlockA.setBlockHash(blockHeaderChainA.getHash());
-    trieLogLayerBlockA.writeTo(rlpLogBlockA);
     final BytesValueRLPOutput rlpLogBlockB = new BytesValueRLPOutput();
     final TrieLogLayer trieLogLayerBlockB = new TrieLogLayer();
     trieLogLayerBlockB.setBlockHash(blockHeaderChainB.getHash());
     trieLogLayerBlockB.writeTo(rlpLogBlockB);
-    when(keyValueStorage.get(blockHeaderChainA.getHash().toArrayUnsafe()))
-        .thenReturn(Optional.of(rlpLogBlockA.encoded().toArrayUnsafe()));
     when(keyValueStorage.get(blockHeaderChainB.getHash().toArrayUnsafe()))
         .thenReturn(Optional.of(rlpLogBlockB.encoded().toArrayUnsafe()));
 

--- a/ethereum/core/src/test/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchiveTest.java
+++ b/ethereum/core/src/test/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchiveTest.java
@@ -21,6 +21,7 @@ import static org.hyperledger.besu.ethereum.bonsai.BonsaiWorldStateKeyValueStora
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.eq;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.spy;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
@@ -199,5 +200,59 @@ public class BonsaiWorldStateArchiveTest {
     verify(layeredWorldStatesByHash).entrySet();
     verify(updater, times(1)).rollBack(any());
     verify(updater, times(1)).rollForward(any());
+  }
+
+  @SuppressWarnings({"unchecked"})
+  @Test
+  public void testGetMutableWithRollbackNotOverrideTrieLogLayer() {
+    final KeyValueStorageTransaction keyValueStorageTransaction =
+        mock(KeyValueStorageTransaction.class);
+    when(keyValueStorage.startTransaction()).thenReturn(keyValueStorageTransaction);
+    final BlockHeader genesis = blockBuilder.number(0).buildHeader();
+    final BlockHeader blockHeaderChainA =
+        blockBuilder.number(1).timestamp(1).parentHash(genesis.getHash()).buildHeader();
+    final BlockHeader blockHeaderChainB =
+        blockBuilder.number(1).timestamp(2).parentHash(genesis.getHash()).buildHeader();
+
+    final Map<Bytes32, BonsaiLayeredWorldState> layeredWorldStatesByHash = mock(HashMap.class);
+    when(layeredWorldStatesByHash.containsKey(any(Bytes32.class))).thenReturn(true);
+    when(layeredWorldStatesByHash.get(eq(blockHeaderChainA.getHash())))
+        .thenReturn(mock(BonsaiLayeredWorldState.class, Answers.RETURNS_MOCKS));
+    when(layeredWorldStatesByHash.get(eq(blockHeaderChainB.getHash())))
+        .thenReturn(mock(BonsaiLayeredWorldState.class, Answers.RETURNS_MOCKS));
+
+    bonsaiWorldStateArchive =
+        spy(new BonsaiWorldStateArchive(storageProvider, blockchain, 12, layeredWorldStatesByHash));
+    var updater = spy(bonsaiWorldStateArchive.getUpdater());
+    when(bonsaiWorldStateArchive.getUpdater()).thenReturn(updater);
+
+    // initial persisted state hash key
+    when(blockchain.getBlockHeader(eq(Hash.ZERO))).thenReturn(Optional.of(blockHeaderChainA));
+    // fake trie log layer
+    final BytesValueRLPOutput rlpLogBlockA = new BytesValueRLPOutput();
+    final TrieLogLayer trieLogLayerBlockA = new TrieLogLayer();
+    trieLogLayerBlockA.setBlockHash(blockHeaderChainA.getHash());
+    trieLogLayerBlockA.writeTo(rlpLogBlockA);
+    final BytesValueRLPOutput rlpLogBlockB = new BytesValueRLPOutput();
+    final TrieLogLayer trieLogLayerBlockB = new TrieLogLayer();
+    trieLogLayerBlockB.setBlockHash(blockHeaderChainB.getHash());
+    trieLogLayerBlockB.writeTo(rlpLogBlockB);
+    when(keyValueStorage.get(blockHeaderChainA.getHash().toArrayUnsafe()))
+        .thenReturn(Optional.of(rlpLogBlockA.encoded().toArrayUnsafe()));
+    when(keyValueStorage.get(blockHeaderChainB.getHash().toArrayUnsafe()))
+        .thenReturn(Optional.of(rlpLogBlockB.encoded().toArrayUnsafe()));
+
+    when(blockchain.getBlockHeader(eq(blockHeaderChainB.getHash())))
+        .thenReturn(Optional.of(blockHeaderChainB));
+    when(blockchain.getBlockHeader(eq(genesis.getHash()))).thenReturn(Optional.of(genesis));
+
+    assertThat(bonsaiWorldStateArchive.getMutable(null, blockHeaderChainB.getHash()))
+        .containsInstanceOf(BonsaiPersistedWorldState.class);
+
+    // verify is not persisting if already present
+    verify(keyValueStorageTransaction, never())
+        .put(eq(blockHeaderChainA.getHash().toArrayUnsafe()), any());
+    verify(keyValueStorageTransaction, never())
+        .put(eq(blockHeaderChainB.getHash().toArrayUnsafe()), any());
   }
 }

